File: src/main/java/com/iminling/core/config/argument/DefaultRequestDataReader.java
package com.iminling.core.config.argument;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpInputMessage;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpRequest;
import org.springframework.http.MediaType;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.web.method.HandlerMethod;

import javax.validation.constraints.NotNull;
import java.io.IOException;
import java.io.InputStream;
import java.io.PushbackInputStream;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

@Configuration
public class DefaultRequestDataReader {

    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultRequestDataReader.class);
    protected static final Set<HttpMethod> SUPPORTED_METHODS = EnumSet.of(HttpMethod.POST, HttpMethod.PUT, HttpMethod.PATCH);

    private final ObjectMapper objectMapper;

    public DefaultRequestDataReader() {
        MappingJackson2HttpMessageConverter mappingJackson2HttpMessageConverter = new MappingJackson2HttpMessageConverter();
        this.objectMapper = mappingJackson2HttpMessageConverter.getObjectMapper();
        this.objectMapper.configure(JsonParser.Feature.ALLOW_COMMENTS, true);
    }

    public boolean canRead(@NotNull HttpInputMessage message) {
        MediaType mediaType = message.getHeaders().getContentType();
        if (!canRead(mediaType)) {
            return false;
        }
        HttpMethod httpMethod = (message instanceof HttpRequest ? ((HttpRequest) message).getMethod() : null);
        return canRead(httpMethod);
    }

    protected boolean canRead(@NotNull MediaType mediaType) {
        if (mediaType == null){
            return true;
        }
        for (MediaType supportedMediaType : getSupportedMediaTypes()) {
            if (supportedMediaType.includes(mediaType)) {
                return true;
            }
        }
        LOGGER.info("Not support mediaType: {}", mediaType);
        return false;
    }

    protected boolean canRead(@NotNull HttpMethod httpMethod) {
        if (httpMethod == null){
            return true;
        }
        if (SUPPORTED_METHODS.contains(httpMethod)) {
            return true;
        }
        LOGGER.debug("Not support request method: {}", httpMethod);
        return false;
    }

    public JsonNode read(HttpInputMessage message, HandlerMethod handlerMethod) throws IOException {
        InputStream inputStream = message.getBody();
        InputStream body;
        if (inputStream.markSupported()){
            inputStream.mark(1);
            body = (inputStream.read() != -1) ? inputStream : null;
        } else {
            PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream);
            int read = pushbackInputStream.read();
            if (read == -1){
                body = null;
            } else {
                body = pushbackInputStream;
                pushbackInputStream.unread(read);
            }
        }
        if (body != null) {
            return objectMapper.readTree(body);
        }
        return null;
    }

    public List<MediaType> getSupportedMediaTypes() {
        return Arrays.asList(MediaType.APPLICATION_JSON_UTF8, MediaType.APPLICATION_JSON, MediaType.TEXT_PLAIN);
    }
}

File: iminling-core/src/main/kotlin/com/iminling/core/config/argument/DefaultRequestDataReader.kt
package com.iminling.core.config.argument

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.http.HttpInputMessage
import org.springframework.http.HttpMethod
import org.springframework.http.HttpRequest
import org.springframework.http.MediaType
import org.springframework.web.method.HandlerMethod
import java.io.IOException
import java.io.InputStream
import java.io.PushbackInputStream
import javax.validation.constraints.NotNull

/**
 * @author  yslao@outlook.com
 * @since  2021/11/28
 */
class DefaultRequestDataReader(val objectMapper: ObjectMapper) {

    private val logger = LoggerFactory.getLogger(DefaultRequestDataReader::class.java)

    private val suppotedMethod = setOf(HttpMethod.POST, HttpMethod.PUT, HttpMethod.PATCH)

    fun canRead(message: @NotNull HttpInputMessage?): Boolean {
        val mediaType = message!!.headers.contentType
        if (!canRead(mediaType)) {
            return false
        }
        val httpMethod = if (message is HttpRequest) (message as HttpRequest).method else null
        return canRead(httpMethod)
    }

    private fun canRead(mediaType: @NotNull MediaType?): Boolean {
        if (mediaType == null) {
            return true
        }
        for (supportedMediaType in getSupportedMediaTypes()) {
            if (supportedMediaType.includes(mediaType)) {
                return true
            }
        }
        logger.info("Not support mediaType: {}", mediaType)
        return false
    }

    private fun canRead(httpMethod: @NotNull HttpMethod?): Boolean {
        if (httpMethod == null) {
            return true
        }
        if (suppotedMethod.contains(httpMethod)) {
            return true
        }
        logger.debug("Not support request method: {}", httpMethod)
        return false
    }

    @Throws(IOException::class)
    fun read(message: HttpInputMessage, handlerMethod: HandlerMethod?): JsonNode? {
        val inputStream = message.body
        val body: InputStream?
        if (inputStream.markSupported()) {
            inputStream.mark(1)
            body = if (inputStream.read() != -1) inputStream else null
        } else {
            val pushbackInputStream = PushbackInputStream(inputStream)
            val read = pushbackInputStream.read()
            if (read == -1) {
                body = null
            } else {
                body = pushbackInputStream
                pushbackInputStream.unread(read)
            }
        }
        return if (body != null) {
            objectMapper.readTree(body)
        } else null
    }

    fun getSupportedMediaTypes(): List<MediaType> {
        return listOf(MediaType.APPLICATION_JSON, MediaType.TEXT_PLAIN)
    }

}